### ISSUE TITLE
Redraw problems on Raspberry (Raspbian, Debian Jessie)

### ISSUE THREAD

**Issue by fhaun** [i3/i3#2316]
Output of `i3 --moreversion 2>&- || i3 --version`:

```
Binary i3 version:  4.12-44-g42a3371 (2016-04-23, branch "next") © 2009 Michael Stapelberg and contributors
Running i3 version: 4.12-44-g42a3371 (2016-04-23, branch "next") (pid 13143)
Loaded i3 config: /home/fhaun/.i3/config (Last modified: Sun 24 Apr 2016 05:18:35 CEST, 28868 seconds ago)

The i3 binary you just called: /home/fhaun/opt/i3-next/usr/bin/i3
The i3 binary you are running: /home/fhaun/opt/i3-next/usr/bin/i3
```

URL to a logfile as per http://i3wm.org/docs/debugging.html:

http://logs.i3wm.org/logs/5693341270278144.bz2

**What I did:**

Running i3 on Raspberry Pi 3, OS is Raspbian (Debian Jessie). I've tried color depth of 16, 24 and 32. With and without an compositor (compton).

**What I saw:**

I have redraw problems in **all programs**, e.g. an Emacs buffer:
http://i.imgur.com/krE7C20.png

This happens after scrolling via mouse or keyboard. Terminal outputs, e.g. while compiling are also not correctly updated.

After a git bisect I found the commit when the redraw problems begins.

```
780cb8d15d07febbb9be47cb1bc1201072b31838 is the first bad commit
commit 780cb8d15d07febbb9be47cb1bc1201072b31838
Author: Ingo Bürk <ingo.buerk@tngtech.com>
Date:   Tue Nov 17 12:50:06 2015 +0100

    Use 32-bit visual by default if available.
    
    With this patch, we use 32-bit visuals per default whenever it is
    available. Otherwise, we fall back to the actual root window's
    depth, which will typically be 24-bit.
    
    Before this patch, we already used 32-bit depth for containers with
    a window that uses 32-bit. However, this means that we didn't use
    32-bit for split parent containers on which decoration is drawn.
    For 32-bit windows using transparency, this caused a graphical glitch
    because the decoration pixmap behind it would show through. This
    behavior is fixed with this change.
    
    relates to #1278

:040000 040000 9cbc0df993e1b5b2aca17718d8956158182eb278 4b053c965aefcf38b2fa15bfe1773f6de9592b6e M      include
:040000 040000 8d61a43db08a1cb50af4d513582949a16a32ef24 da95393dd0c366982115bcdbf07fdc15f735c4bd M      libi3
:040000 040000 4d9dd355c7e8bf67f4aad5ee276108f60b3d86af c28a12887d0506c3581201e2280278d397eb7b28 M      src
```

**Comment by Airblader** [comment]
For reference, this originated here: https://www.reddit.com/r/i3wm/comments/4g51zc/i3_version_412_redraw_problems_on_raspberry/

**Comment by stapelberg** [comment]
If we don’t find another solution for this issue, we should revert the commit. AFAIR, it was only introduced for transparency, which is not an officially supported use-case, right?

**Comment by Airblader** [comment]
I don't think we should revert the commit without understanding what is happening. I am not aware of any reason for why we can't use these depths, so I'm not convinced it's a bug in i3. It might just be triggering a code path somewhere that wasn't triggered before. It doesn't seem to be a general issue because I'm pretty sure we have quite a few more emacs users and 4.12 has been out for a while now. Plus we are dealing with "exotic" hardware here.

**Comment by Airblader** [comment]
Since the referenced patch was only one of a larger change and has been partially reverted already, I think the best course of action is to manually write a revert patch that @fhaun can apply so we can gain some confidence in what exactly is causing these issues.

@fhaun Can you also provide the output of both of these commands, clicking on the emacs window when executing them?

    xwininfo | grep -i depth
    xwininfo -frame | grep -i depth

**Comment by Airblader** [comment]
@fhaun And then, please try appending the following two patches (individually, not together) and let us know if either one fixes the problem.

#### Patch 1: Don't clear the surface of the first child

````
diff --git a/src/x.c b/src/x.c
index 0fbc4d1..7ca1f03 100644
--- a/src/x.c
+++ b/src/x.c
@@ -520,14 +520,6 @@ void x_draw_decoration(Con *con) {
     if (parent->frame_buffer.id == XCB_NONE)
         goto copy_pixmaps;
 
-    /* For the first child, we clear the parent pixmap to ensure there's no
-     * garbage left on there. This is important to avoid tearing when using
-     * transparency. */
-    if (con == TAILQ_FIRST(&(con->parent->nodes_head))) {
-        draw_util_clear_surface(conn, &(con->parent->frame_buffer), COLOR_TRANSPARENT);
-        FREE(con->parent->deco_render_params);
-    }
-
     /* 4: paint the bar */
     draw_util_rectangle(conn, &(parent->frame_buffer), p->color->background,
                         con->deco_rect.x, con->deco_rect.y, con->deco_rect.width, con->deco_rect.height);

```

#### Patch 2: Don't try to use 32-bit depth by default

````
diff --git a/src/main.c b/src/main.c
index b2ce17d..2aef598 100644
--- a/src/main.c
+++ b/src/main.c
@@ -483,25 +483,7 @@ int main(int argc, char *argv[]) {
 
     root_depth = root_screen->root_depth;
     colormap = root_screen->default_colormap;
-    visual_type = xcb_aux_find_visual_by_attrs(root_screen, -1, 32);
-    if (visual_type != NULL) {
-        root_depth = xcb_aux_get_depth_of_visual(root_screen, visual_type->visual_id);
-        colormap = xcb_generate_id(conn);
-
-        xcb_void_cookie_t cm_cookie = xcb_create_colormap_checked(conn,
-                                                                  XCB_COLORMAP_ALLOC_NONE,
-                                                                  colormap,
-                                                                  root,
-                                                                  visual_type->visual_id);
-
-        xcb_generic_error_t *error = xcb_request_check(conn, cm_cookie);
-        if (error != NULL) {
-            ELOG("Could not create colormap. Error code: %d\n", error->error_code);
-            exit(EXIT_FAILURE);
-        }
-    } else {
-        visual_type = get_visualtype(root_screen);
-    }
+    visual_type = get_visualtype(root_screen);
 
     DLOG("root_depth = %d, visual_id = 0x%08x.\n", root_depth, visual_type->visual_id);
     DLOG("root_screen->height_in_pixels = %d, root_screen->height_in_millimeters = %d, dpi = %d\n",

```

**Comment by fhaun** [comment]
xwininfo before patching:
```
$ xwininfo | grep -i depth
Depth: 24

$ xwininfo -frame | grep -i depth
Depth: 32
```

"Patch 1: Don't clear the surface of the first child" doesn't help.

"Patch 2: Don't try to use 32-bit depth by default" **solves** the redraw problems.

xwininfo after patching with Patch 1:
```
$ xwininfo | grep -i depth
Depth: 24

$ xwininfo -frame | grep -i depth
Depth: 32
```

xwininfo after patching with Patch 2:
```
$ xwininfo | grep -i depth
Depth: 24

$ xwininfo -frame | grep -i depth
Depth: 24
```

With Patch 2 both depth are the same now.

**Comment by Airblader** [comment]
Thanks for the feedback!

@psychon You know your way around the whole thing with visuals and depths. Is there anything wrong about using a 32-bit depth frame window for a 24-bit client window? My understanding is that this is fine – we can't control a client's depth, afterall, but that can't force the window manager to give up transparency on its decorations.

From what I can tell awesome does the same, too. I'm still leaning towards this not being an i3 issue.

**Comment by psychon** [comment]
This is definitely not an i3 issue. It just hits a different code path in the X server (if both windows have the same depth, the server uses the same memory for storing the content of both, if they have different depths, then it does $SOMETHING to keep them somewhat apart). I don't know the details, but I know that when awesome started using a 32 bit visual, I had similar problems with the intel driver. These vanished eventually. Then someone reported a bug about slow rendering. Then Chrome with GPU-compositing enabled started being drawn even if it was not on a the visible workspace etc. Most of these issues I could never reproduce.

In the end, awesome gained a `--no-argb` command line option and whenever a bug like this comes in, I ask if this flag makes the issue go away. If so, it's definitely a bug in the server. See the list at https://github.com/awesomeWM/awesome/issues?utf8=%E2%9C%93&q=label%3Ano-argb.

**Comment by Airblader** [comment]
@psychon Thanks :-)

@stapelberg Now we need to decide what to do. Here are the options I see:

* We close the bug and tell people to update their X server. While I can see that this is probably the most uncomfortable for a user, it's my preference. We shouldn't work around bugs in other applications (including the server) unless it doesn't restrict us (as was the case for having to clear a cairo surface twice). That's the argument we use when client applications have bugs as well and makes sense to me.
* We revert the commit. For the reasons above, I highly object this. :-) If we do go for this I'd at least like to do it with a `#if 0` rather than removing it altogether.
* We introduce a command-line option similar to awesome. Would be fine with me, but my feeling is that you wouldn't like that.
* We introduce a command-line option but with inverted logic (disabled by default). Also fine with me.

**Comment by stapelberg** [comment]
Currently I’m leaning towards reverting the commit. That avoids a bunch of bugs (in otherwise supported configurations).

Can you clarify if anything aside from transparency-related issues would regress when we revert the commit?

**Comment by Airblader** [comment]
> (in otherwise supported configurations).

Why is a machine running a broken X server a supported configuration? If we had this change for two years rather than a few months and the X server only then introduced the bug, would we also revert commits for that?

I think that we should keep the pressure on both Xorg and Debian to fix bugs and backport fixes. If everybody just ducks away, there's no reason to fix these problems properly, but at the same time it paralyzes development in applications. I think the *proper* way to deal with bugs is fixing it in the application that has the bug and that's the stance we took in the past as well.

Also, we have already released this change. And while we don't support transparency and can happily close any report coming in, it would probably still upset users. On the other hand, over the course of almost two months since 4.12, we've only had a single person reporting this issue.

**Comment by Airblader** [comment]
@psychon Are you aware of any existing bug reports against X for this? I'm just wondering why awesome has had this for so long and yet the bug seems to persist.

@stapelberg If my philosophy on bug responsibility above doesn't convince you, can we agree on the `#if 0` approach with an explanatory comment? Or maybe even the command-line flag?

**Comment by psychon** [comment]
@Airblader Nope. I've been too lazy about sending bug reports upstream, mostly because too many users use the argument "but it works in fluxbox" as a reason why I have to fix this. Also, https://bugs.freedesktop.org/show_bug.cgi?id=34885 didn't really have any nice result (and it took ages to come up with a simple, self-contained reproducer; this is even harder for bugs like these which I cannot reproduce with my ancient intel chip). I guess I gave up too much on this...
(Also: I don't know the details, but doesn't raspbian have some closed-source graphics driver?)

**Comment by Airblader** [comment]
> (Also: I don't know the details, but doesn't raspbian have some closed-source graphics driver?)

If it does, we should reject this bug anyway. We don't support closed-source software. :-)

**Comment by Airblader** [comment]
Just to clarify, that last comment wasn't meant to be serious. ;)

**Comment by fhaun** [comment]
Something about the graphics driver:
http://liliputing.com/2014/02/raspberry-pi-gets-true-open-source-graphics-drivers.html

**Comment by stapelberg** [comment]
By “supported configuration” I mean the Raspberry Pi 3 — I’m not sure where the bug is precisely. If it’s a bug that can be worked around by upgrading (individual packages or the entire operating system) from Debian stable to Debian testing, I’m fine with not addressing it in i3 itself. However, if the bug isn’t reported/accepted upstream, and prevents all users on a Raspberry Pi 3 from using i3, I think we should do something about it.

**Comment by Airblader** [comment]
OK, fair enough. So far we only have a single Pi3 user reporting this. We'll see if others report it too. 

@fhaun Would you be willing to upgrade your system to see whether it changes anything?

**Comment by mborejdo** [comment]
looking on the provided screenshot, I also experience these kind of rendering-artifacts.

For me this happens often on chromium.

I experience these problems connecting to a running i3-wm through Team-Viewer.
I am running I3 inside a docker-container. So some more moving parts here.

My setup:
`
$ i3 --moreversion 2>&- || i3 --version:
Binary i3 version:  4.12-62-g1572004 (2016-04-26, branch "gaps-next") © 2009 Michael Stapelberg and contributors
Running i3 version: 4.12-62-g1572004 (2016-04-26, branch "gaps-next") (pid 93)
`

TeamViewer (on Docker-host)
`
$ teamviewer --version
 TeamViewer                           11.0.52520  (DEB) 
`

(Docker-)Host:
`
Linux mib.docker.em.lan 4.2.0-35-generic #40-Ubuntu SMP Tue Mar 15 22:15:45 UTC 2016 x86_64 GNU/Linux
`

I3 is running inside a debian stretch based container.

**Comment by Airblader** [comment]
@mborejdo Rendering artifacts all in itself aren't unusual. The standard question is: are you running a compositor? If not, does running one help? And if it doesn't (as was the case here), does the mentioned patch above fix the issue?

**Comment by mborejdo** [comment]
@Airblader i am running compton. I havent had the chance to apply the patches yet. Will report back once I have.

**Comment by Airblader** [comment]
> i am running compton.

I know this sounds wonky, but then the reverse applies: what if you don't run a compositor? The reason being that compton has many settings, a lot of which can cause tears and glitches of all kinds.

**Comment by mborejdo** [comment]
@Airblader I can confirm these "artifacts" still remain after killing compton. I didnt have a chance to patch yet, will soon.

**Comment by fhaun** [comment]
@Airblader Raspbian is up do date.

OK, I've testet i3 with Arch Linux on the Raspberry 3. Everything is fine now. Arch comes with "i3 version  4.12 (2016-03-06, branch "4.12")". No redraw problems and i3's 16-bit color support works like a charm.

**Comment by fhaun** [comment]
@Airblader Raspbian (stable, Jessie) is up do date.

OK, I've testet i3 with Arch Linux on the Raspberry 3. Everything is fine now. Arch comes with "i3 version 4.12 (2016-03-06, branch "4.12")". No redraw problems and i3's 16-bit color support works like a charm.

**Comment by Airblader** [comment]
The fact that an Arch install works now shows even more that this is an issue with something Debian ships – and hence something we shouldn't address by removing code in i3.

**Comment by fhaun** [comment]
Bad news. Redraw (looks more as an window update) problems still exists. Not in Emacs, but in Firexfox and xterm. With "Patch 2: Don't try to use 32-bit depth by default" i3 works much better here.
